doc/dev/haskell-idioms.md

```diff
@@ -0,0 +1,182 @@
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Some f where
+  Some :: f x -> Some f  -- the type does not mention `x`
+```
+(`Some` is canonically available as [`Data.Parameterized.Some.Some`].) To use
+a `Some`, you generally pattern match on it. This brings the captured variable
+into scope (though just as an unconstrained variable, so any code you write
+after the pattern match has to be polymorphic over it).
+```haskell
+wasItJust :: Some Maybe -> String
+wasItJust someMaybe =
+  case someMaybe of
+    -- Can use _x here, but have to be completely polymorphic over its type. Not
+    -- really much you *can* do with a value of an unknown type.
+    Some (Just _x) -> "it was Just"
+    Some Nothing -> "it was Nothing"
+
+main :: IO ()
+main = putStrLn (wasItJust (Some (Just "ok")))
+```
+Sometimes, we want to bring the variable into scope by pattern-matching on
+the constructor, but we'd rather not use a `case` statement, as it increases
+indentation (e.g., we might be pattern-matching on several `Some`s in a row). We
+can't do this with `let`, because the type variable is only in scope for the RHS
+of the `=`.
+```haskell
+wontWork :: Some Maybe -> IO ()
+wontWork someMaybe = do
+  let Some x = someMaybe
+  -- try something with x...
+  pure ()
+```
+Not only will this not compile, it'll give us a confusing error message:
+```
+Couldn't match expected type ‘p0’ with actual type ‘Maybe x’
+        because type variable ‘x’ would escape its scope
+      This (rigid, skolem) type variable is bound by
+        a pattern with constructor:
+        ...
+```
+However, this *is* possible using `do`-notation. Since our `Some` is just a
+pure value (not a monadic action), we have to "inject" it into the monad using
+`pure`:
+```haskell
+printWasItJust :: Some Maybe -> IO ()
+printWasItJust someMaybe = do
+  Some x <- pure someMaybe
+  -- Again, can use x here, but only polymorphically
+  when (isJust x) $
+    putStrLn "it was Just"
+```
+To understand why this works, desugar the `do` notation:
+```haskell
+printWasItJust :: Some Maybe -> IO ()
+printWasItJust someMaybe =
+  pure someMaybe >>= \(Some x) -> when (isJust x) (putStrLn "it was Just")
+```
+That's why you might see `do pat <- pure val`.
+
+[`Data.Parameterized.Some.Some`]: https://hackage-content.haskell.org/package/parameterized-utils-2.1.10.0/docs/Data-Parameterized-Some.html#t:Some
+
+## Matching on constraint-capturing constructors in `do` blocks
+
+[GADT] constructors can "capture" constraints, which can be recovered by
+pattern-matching on them. Note the lack of a `Show a` constraint on `showIt` in
+the following example:
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Showable a where
+  Showable :: Show a => a -> Showable a
+
+showIt :: Showable a -> String
+showIt (Showable x) = show x  -- pattern matching brings Show a into scope
+
+main :: IO ()
+main = putStrLn (showIt (Showable "hello"))
+```
+(Operationally, the GADT constructor holds the vtable for the class.) GHC's
```

Comment:
"vtable" is jargon, and one that I'm not sure if most Haskell programmers will know. We could consider "dictionary" instead, although that is also jargon. Perhaps it would be worth leaving a link that defines one of these two terms?